### PR TITLE
Update CI in OpenMPTarget to use llvm-17 

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR updates OpenMPTarget CI to use llvm/17.0.1 release. 
llvm/17.0.0 release is retracted due to issues. [](https://github.com/llvm/llvm-project/releases/tag/llvmorg-17.0.1)

The PR also updates clang version in OpenMPTarget CI to use cmake/3.20.0 which is the new minimum cmake version to build clang. [](https://llvm.org/docs/CMake.html#quick-start)

### CODE REVIEW

File: algorithms/unit_tests/CMakeLists.txt [hunk 1]
```diff
@@ -410,6 +410,13 @@ if(KOKKOS_ENABLE_OPENMPTARGET)
   )
 endif()
 
+# FIXME_OPENMPTARGET - remove sort test as it leads to ICE with clang/17 and above at compile time.
+if(KOKKOS_ENABLE_OPENMPTARGET STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER 16.0.0)
```
Comment (dalg24) [inline algorithms/unit_tests/CMakeLists.txt:414]: Shouldn't it be `VERSION_GREATER_EQUAL 17.0.0 ` in that case?

Comment (rgayatri23) [reply algorithms/unit_tests/CMakeLists.txt]: Oh yes you are right. Fixed it

**Comment by masterleinad** [comment]
```
4: [ RUN      ] openmptarget.reducers_double
4: "PluginInterface" error: Faliure to copy data from device to host. Pointers: host = 0x00007fffffffd190, device = 0x00007fffba002000, size = 24: Error in cuMemcpyDtoHAsync: misaligned address
4: Libomptarget error: Copying data from device failed.
4: Libomptarget error: Call to targetDataEnd failed, abort target.
4: Libomptarget error: Failed to process data after launching the kernel.
4: Libomptarget error: Consult https://openmp.llvm.org/design/Runtimes.html for debugging options.
4: Kokkos_OpenMPTarget_Parallel_Common.hpp:111:1: Libomptarget fatal error 1: failure of target construct while offloading is mandatory
 4/46 Test  #4: Kokkos_CoreUnitTest_OpenMPTarget ............Subprocess aborted***Exception:  73.45 sec
```

**Review by masterleinad — COMMENTED** [review]
It looks strange that you are (still) only disabling tests if the minor or patch version is greater than 0. I would expect, e.g., `16.0.0` to have more bugs than `16.0.1` or `16.1.0`.
Do you mean "greater or equal" rather than "greater" in these cases? Or do you mean, e.g., `>= 17` instead of `>16` in some cases?

File: scripts/docker/Dockerfile.openmptarget [hunk 2]
```diff
@@ -23,7 +23,7 @@ RUN KEYDUMP_URL=https://cloud.cees.ornl.gov/download && \
     gpg --verify ${KEYDUMP_FILE}.sig ${KEYDUMP_FILE} && \
     rm ${KEYDUMP_FILE}*
 
-ARG CMAKE_VERSION=3.18.5
+ARG CMAKE_VERSION=3.20.0
```
Comment (dalg24) [inline scripts/docker/Dockerfile.openmptarget:26]: Go for the latest patch release (v3.20.6 I think)

**Comment by rgayatri23** [comment]
Restart this please

**Comment by rgayatri23** [comment]
Retest this please

File: core/unit_test/CMakeLists.txt [hunk 3]
```diff
@@ -362,13 +365,28 @@ foreach(PairDeviceSpace HIP-HostPinned;HIP-Managed;Cuda-HostPinned;Cuda-UVM;SYCL
...
 
-
+# Disable non-compiling tests based on clang version.
 if(Kokkos_ENABLE_OPENMPTARGET)
   list(REMOVE_ITEM OpenMPTarget_SOURCES
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_Other.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamCombinedReducers.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamReductionScan.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_WorkGraph.cpp
+    IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER 15.0.0)
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:375]: should this be VERSION_GREATER_EQUAL_16?

[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:375]: I just assume it does

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
The OpenMPTarget build passed.  Ignoring the rest.

**Pull request merged**